File: 2023/day02/day02.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func get_draws(line string) [][]string {
	var draws [][]string

	for _, draw := range strings.Split(strings.Split(line, ":")[1], ";") {
		var draw_list []string
		for _, count := range strings.Split(draw, ", ") {
			draw_list = append(draw_list, strings.ReplaceAll(strings.TrimSpace(count), "\n", ""))
		}
		draws = append(draws, draw_list)
	}

	return draws
}

func get_maximum_color_counts(draws [][]string) []int {
	maximum_values := map[string]int{
		"red":   0,
		"green": 0,
		"blue":  0,
	}

	for _, draw := range draws {
		for _, count := range draw {
			split_count := strings.Split(string(count), " ")
			value, _ := strconv.Atoi(split_count[0])
			color := split_count[1]

			if value > maximum_values[color] {
				maximum_values[color] = value
			}
		}
	}

	var values []int
	for _, value := range maximum_values {
		values = append(values, value)
	}

	return values
}

func get_product(counts []int) int {
	product := 1
	for _, number := range counts {
		product *= number
	}
	return product
}

func main() {
	file, _ := os.ReadFile("input.txt")

	total := 0
	for _, line := range strings.Split(string(file), "\n") {
		total += get_product(get_maximum_color_counts(get_draws(strings.TrimSpace(line))))
	}

	fmt.Println(total)
}

File: 2023/day02/day02.py
from functools import reduce
import operator


def get_draws(line: str) -> list[list[str]]:
    draws = []

    for draw in line.split(": ")[1].split("; "):
        draws.append([count.replace("\n", "") for count in draw.split(", ")])

    return draws


def get_maximum_color_counts(draws: list[list[str]]) -> list[int]:
    maximum_values = {
        "red": 0,
        "green": 0,
        "blue": 0,
    }

    for draw in draws:
        for count in draw:
            value, color = count.split()
            if int(value) > maximum_values[color]:
                maximum_values[color] = int(value)

    return maximum_values.values()


if __name__ == "__main__":

    with open("input.txt") as f:
        lines = f.readlines()

    total = 0
    for line in lines:
        total += reduce(operator.mul, get_maximum_color_counts(get_draws(line)), 1)

    print(total)
